bookserver/crud.py

```diff
@@ -115,22 +115,21 @@ async def fetch_last_answer_table_entry(
 
 # Courses
 # -------
-async def fetch_course(course_name: str) -> CoursesValidator:
-    query = select(Courses).where(Courses.course_name == course_name)
+async def fetch_base_course(base_course: str) -> CoursesValidator:
+    query = select(Courses).where(Courses.base_course == base_course)
     async with async_session() as session:
         res = await session.execute(query)
-    # When selecting ORM entries it is useful to use the ``scalars`` method
-    # This modifies the result so that you are getting the ORM object
-    # instead of a Row object. `See <https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-orm-entities-and-attributes>`_
-    return CoursesValidator.from_orm(res.scalars().first())
+        # When selecting ORM entries it is useful to use the ``scalars`` method
+        # This modifies the result so that you are getting the ORM object
+        # instead of a Row object. `See <https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-orm-entities-and-attributes>`_
+        course = res.scalars().one_or_none()
+        return CoursesValidator.from_orm(course) if course else None
 
 
-async def create_course(course_info: CoursesValidator) -> CoursesValidator:
+async def create_course(course_info: CoursesValidator):
```

Comment:
There are many use cases where it is useful to get back the thing you just added.  Why return None?  At least maybe a boolean to indicate success / failure.

Reply:
For returning a boolean -- how do we want to handle exceptions/failures for crud operations? Trap the exception and return something nice, or allow the exception to terminate the operation? Or add another exception handler class for database exceptions that will return something besides a 501 plus log the error?

We could certainly return exactly what was passed to the function. I haven't needed that in my code, but I'm happy to provide it if you think it's useful.

Reply:
The difference is that the returned value would have the id, which is what is needed in many cases.

One thing I do like about web2py is the tickets it generates. (well I don't like the raw form of the tickets) but the idea that whenever there is a crash it stores the crash in the errors folder so that I can investigate.

We could do that with an exception handler middleware that would also return the stack trace as part of the javascript response.  Rather than just a 501 return code.
over time this would allow our client side code to be more user friendly in providing the users the detail they need to report errors.

Reply:
I like the idea of middleware to capture exceptions and let us know!

I'll add returns of the created objects.

Reply:
Thanks, just as an example, logBookEvent does check the value returned from `create_useinfo_entry` to determine what to return to the client.  Its not a great example as I don't know that anything checks the value returned by logBookEvent in the browser, but still....

Reply:
My next PR will have everything in rslogger.py returning a `JSONResponse` object.